### PR TITLE
cgroup: wait to verify value in set()

### PR DESCRIPTION
**Pull request by setrofim**
This is a fix for https://github.com/ARM-software/devlib/issues/136

The standard behavior for write_value() is to read it back immediately
afterwards and verify that it has been written. When writing some cgroup
parameter values (e.g. "FROZEN"), it may take a while for them to take
effect. If they are read back immediately afterwards, the result won't
match the expected value, producing an error.

While it is possible to disable validation on write, it is still
desirable to make sure that the parameters are set properly. This commit
changes the behavior to continue reading the value up to a timeout
period afterwards, giving it time to take effect, and only raise an
error if the timeout is reached.

### CODE REVIEW

**Comment by qperret** [comment]
Thanks for the fix. 

Please `import os.path`. Apart from that it looks good to me. It's probably a shame to wait before raising an exception if the read value if clearly wrong but I don't really see how to do that in a generic way. This is probably the best we can do without managing all transition states of cgroups independently.

**Comment by setrofim** [comment]
> Please `import os.path`.

Done.

>  It's probably a shame to wait before raising an exception if the read value if clearly wrong but I don't really see how to do that in a generic way. 

Well this is only in the exceptional case, and only 10s of target time. The alternative would be to add an awareness of in-progress/final value pairs, such as FREEZING/FROZEN, but that would periodically break when things change.

**Comment by qperret** [comment]
> The alternative would be to add an awareness of in-progress/final value pairs, such as FREEZING/FROZEN, but that would periodically break when things change.

I agree, it's probably not worth it.

Design-wise, would it make sense to move the timeout feature to `target.write_value` ? It could be used together with `verify` to wait before raising an exception. I would suggest a default timeout of -1 meaning raise exception with no delay.

File: devlib/module/cgroups.py [hunk 1]
```diff
@@ -207,13 +208,15 @@ def tasks_per_group(self):
             tasks[cg] = self.tasks_count(cg)
         return tasks
 
+
 class CGroup(object):
 
-    def __init__(self, controller, name, create=True):
+    def __init__(self, controller, name, create=True, set_timeout=10):
```
Comment (derkling) [inline devlib/module/cgroups.py]: The timeout is really required only for the `freezer` controller... it seems a bit overfilling to have general timeout defined here for each and every controller.

What about moving it to the `set(self, **attrs)` method below or, maybe even better, fix the only really broken client, which is the `freeze` method defined at the bottom of this module?
Maybe we can implement freeze/thaw operations as a dedicated shutil call, which will properly include the required timeout for this operation only, to be used by the freeze() method.

**Comment by derkling** [comment]
I would better consider the shutil alternative implementation to better support freeze/thaw operations.

**Comment by setrofim** [comment]
> I would better consider the shutil alternative implementation to better support freeze/thaw operations.

Yeah, I think that might be a better idea. Closing this pull.

**Pull request closed without merging**